File: src/test/java/ukalus/flood/floodfill/FloodFillTest.java
package ukalus.flood.floodfill;

import ukalus.level.Level;
import ukalus.math.Vector2D;

import java.util.Iterator;
import java.util.Set;

import junit.framework.TestCase;

public class FloodFillTest extends TestCase {

  private static final Object FLOOR = new Object();
  private static final Object WALL = null;

  public void testFill1() {
    Object[][] data = new Object[][]{
        {FLOOR, FLOOR, FLOOR},
        {FLOOR, FLOOR, FLOOR},
        {FLOOR, FLOOR, FLOOR}};
    FloodFill fill = new FloodFill();
    Set<Node> actual = fill.getTemplate(1, new Node(new Level<>(data),
      Vector2D.Companion.get(1, 1), 0));
    assertEquals(9, actual.size());
  }

  public void testFill1WithObstacles() {
    Object[][] data = new Object[][]{
        {FLOOR, WALL, FLOOR},
        {FLOOR, FLOOR, FLOOR},
        {FLOOR, WALL, FLOOR}};
    FloodFill fill = new FloodFill();
    Set<Node> actual = fill.getTemplate(1, new Node(new Level<>(data),
      Vector2D.Companion.get(1, 1), 0));
    assertEquals(7, actual.size());
  }

  public void testFill1WithBlock() {
    Object[][] data = new Object[][]{
        {FLOOR, WALL, FLOOR, FLOOR},
        {FLOOR, FLOOR, FLOOR, FLOOR},
        {FLOOR, FLOOR, WALL, FLOOR},
        {FLOOR, FLOOR, FLOOR, FLOOR},
        {FLOOR, WALL, FLOOR, FLOOR}};
    FloodFill fill = new FloodFill();
    Set<Node> actual = fill.getTemplate(10, new Node(new Level<>(data),
      Vector2D.Companion.get(2, 0), 0));
    assertEquals(17, actual.size());

    Iterator<Node> iterator = actual.iterator();
    Node target = null;
    while (iterator.hasNext() && target == null) {
      Node current = iterator.next();
      if (current.getLocation()
        .equals(Vector2D.Companion.get(2, 3))) {
        target = current;
      }
    }

    assertNotNull(target);
    assertEquals(3, target.getDistance());
  }
}

File: src/test/java/ukalus/flood/floodfill/FloodFillTest.kt
package ukalus.flood.floodfill

import junit.framework.TestCase
import ukalus.level.Level
import ukalus.math.Vector2D

class FloodFillTest : TestCase() {

    fun testFill1() {
        val data = arrayOf(arrayOf(Node.FLOOR, Node.FLOOR, Node.FLOOR), arrayOf(Node.FLOOR, Node.FLOOR, Node.FLOOR), arrayOf(Node.FLOOR, Node.FLOOR, Node.FLOOR))
        val actual = FloodFill().getTemplate(1, Node(Level(data), Vector2D(1, 1), 0))
        assertEquals(9, actual.size)
    }

    fun testFill1WithObstacles() {
        val data = arrayOf(arrayOf(Node.FLOOR, Node.WALL, Node.FLOOR), arrayOf(Node.FLOOR, Node.FLOOR, Node.FLOOR), arrayOf(Node.FLOOR, Node.WALL, Node.FLOOR))
        val actual = FloodFill().getTemplate(1, Node(Level(data), Vector2D(1, 1), 0))
        assertEquals(7, actual.size)
    }

    fun testFill1WithBlock() {
        val data = arrayOf(arrayOf(Node.FLOOR, Node.WALL, Node.FLOOR, Node.FLOOR), arrayOf(Node.FLOOR, Node.FLOOR, Node.FLOOR, Node.FLOOR), arrayOf(Node.FLOOR, Node.FLOOR, Node.WALL, Node.FLOOR), arrayOf(Node.FLOOR, Node.FLOOR, Node.FLOOR, Node.FLOOR), arrayOf(Node.FLOOR, Node.WALL, Node.FLOOR, Node.FLOOR))
        val actual = FloodFill().getTemplate(10, Node(Level(data), Vector2D(2, 0), 0))
        assertEquals(17, actual.size)

        val target = actual.first { it.location == Vector2D(2, 3) }

        assertNotNull(target)
        assertEquals(3, target.distance)
    }


}
